CONTRIBUTING.md

```diff
@@ -0,0 +1,57 @@
+`mvn install -Djavacpp.platform.extension=-mkl` or CUDA with `mvn install -Djavacpp.platform.extension=-gpu`
+or both with `mvn install -Djavacpp.platform.extension=-mkl-gpu`.
+
+When building this project for the first time in a given workspace, the script will attempt to download
+the [TensorFlow runtime library sources](https://github.com/tensorflow/tensorflow) and build of all the native code
+for your platform. This requires a valid environment for building TensorFlow, including the [bazel](https://bazel.build/)
+build tool and a few Python dependencies (please read [TensorFlow documentation](https://www.tensorflow.org/install/source)
+for more details).
+
+This step can take multiple hours on a regular laptop. It is possible though to skip completely the native build if you are
+working on a version that already has pre-compiled native artifacts for your platform [available on Sonatype OSS Nexus repository](#Snapshots).
+You just need to activate the `dev` profile in your Maven command to use those artifacts instead of building them from scratch
+(e.g. `mvn install -Pdev`).
+
+Note that modifying any source files under `tensorflow-core` may impact the low-level TensorFlow bindings, in which case a
+complete build could be required to reflect the changes.
+
+
+## Running Tests
+
+`ndarray` can be tested using the maven `test` target.  `tensorflow-core` and `tensorflow-framework`, however, 
+should be tested using the `integration-test` target, due to the need to include native binaries.
+It will **not** be ran when using the `test` target of parent projects, but will be ran by `install` or `integration-test`.
+
+## Contributing
+
+### Formatting
+
+Java sources should be formatted according to the [Google style guide](https://google.github.io/styleguide/javaguide.html).
+It can be included in [IntelliJ](https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml) and 
+[Eclipse](https://github.com/google/styleguide/blob/gh-pages/eclipse-java-google-style.xml).
+[Google's C++ style guide](https://google.github.io/styleguide/cppguide.html) should also be used for C++ code.
+
+### Code generation
+
+Code generation for `Ops` and related classes is done during `tensorflow-core-api`'s `install`, using the annotation processor in 
+`tensorflow-core-generator`. If you change or add any operator classes (annotated with `org.tensorflow.op.annotation.Operator`), 
+endpoint methods (annotated with `org.tensorflow.op.annotation.Endpoint`), or change the annotation processor, be sure to re-run a 
+full `mvn install` in `tensorflow-core-api`.
+
+### Working with Bazel generation
+
+`tensorflow-core-api` uses Bazel-built C++ code generation to generate most of the `@Operator` classes.  To get it to build, you will likely need to 
+clone the [tensorflow](https://github.com/tensorflow/tensorflow) project, run its configuration script (`./configure`), and copy the resulting 
+`.tf_configure.bazelrc` to `tensorflow-core-api`.
```

Comment:
Why this extra step? I never had to do it before. 

We have already a [`bazelrc`](https://github.com/tensorflow/java/blob/master/tensorflow-core/tensorflow-core-api/.bazelrc) configuration file saved in our repo, I've noticed that it has not been updated after each TF version upgrade like it should, maybe that's the issue?

Reply:
Maybe?  But if you have to do something like enable CUDA, or point it to non-default CUDA dirs, or use a different Python binary (that was my issue, I needed to point it to my conda install), you need the extra configuration.  It's handled in `build.sh` for maven builds, but if you aren't going through build.sh I had issues.  I had continual issues with needing `--incompatible_restrict_string_escapes=false` as well.

Reply:
It was also necessary to get my bazel IDE plugin working, there were times when `bazel build` would work but the plugin's command wouldn't.  I can move it to more of a "if your build doesn't work try this" section.

Reply:
Yes, or something like: "The bazel build is configured for TF Java [continuous integration](link to CI workflow). If your current local setup is different, you might need to reconfigure bazel as follow:..."